libtenzir/src/detail/loader_saver_resolver.cpp

```diff
@@ -0,0 +1,229 @@
+    return !scheme.inner.empty();
+  }
+};
+
+template <typename Plugin>
+auto find_plugin_by_scheme(std::string_view scheme) -> const Plugin* {
+  for (auto&& plugin : plugins::get<Plugin>())
+    if (plugin->supported_uri_scheme() == scheme)
+      return plugin;
+  return nullptr;
+}
+
+template <typename Plugin>
+auto try_plugin_by_uri(const located<std::string>& src)
+  -> try_plugin_by_uri_result<Plugin> {
+  // Not using caf::uri for anything else but checking for URI validity
+  // This is because it makes the interaction with located<...> very difficult
+  if (!caf::uri::can_parse(src.inner))
+    return {};
+  // In a valid URI, the first ':' is guaranteed to separate the scheme from
+  // the rest
+  TENZIR_ASSERT(src.inner.find(':') != std::string::npos);
+  try_plugin_by_uri_result<Plugin> result{};
+  auto scheme_len = src.inner.find(':');
+  result.scheme = make_located_string_view(src, 0, scheme_len);
+  auto non_scheme_offset = scheme_len + 1;
+  if (src.inner.substr(non_scheme_offset).starts_with("//"))
+    // If the URI has an `authority` component, it starts with "//"
+    // We need to skip that before forwarding it to the loader
+    non_scheme_offset += 2;
+  result.non_scheme = make_located_string_view(src, non_scheme_offset);
+  result.plugin = find_plugin_by_scheme<Plugin>(result.scheme.inner);
+  return result;
+}
+
+template <typename ParserPlugin, typename Plugin, typename Parse>
+auto resolve_impl(parser_interface& parser, const located<std::string>& name,
+                  Parse&& parse)
+  -> std::pair<std::unique_ptr<Plugin>, located<std::string_view>> {
+  if (auto uri = try_plugin_by_uri<ParserPlugin>(name); uri.plugin_found()) {
+    // Valid URI, with a matching plugin name
+    if (uri.scheme.inner == uri.plugin->name()) {
+      // URI scheme is identical to the plugin name:
+      // Pass the URI along _without_ the scheme
+      // This is to support URIs like file://foo,
+      // which will be transformed to `file foo`
+      auto r = prepend_token{uri.non_scheme, parser};
+      auto parsed = parse(uri.plugin, r);
+      TENZIR_DIAG_ASSERT(r.at_end());
+      return {std::move(parsed), make_located_string_view(uri.scheme)};
+    }
+    // URI scheme is different from the plugin name:
+    // Assume that the URI scheme is special, and needs to be passed along
+    // This is to support the `gcs`-connector, which needs
+    // the `gs://` scheme
+    auto r = prepend_token{name, parser};
+    auto parsed = parse(uri.plugin, r);
+    TENZIR_DIAG_ASSERT(r.at_end());
+    return {std::move(parsed), make_located_string_view(uri.scheme)};
+  } else if (uri.valid_uri_parsed()) {
+    // Valid URI, but no matching plugin name -> error
+    return {nullptr, make_located_string_view(uri.scheme)};
+  }
+  // Check if `name` could be a valid plugin name
+  auto plugin_name_pattern = pattern::make("[A-Za-z0-9-_]*");
+  TENZIR_DIAG_ASSERT(plugin_name_pattern);
+  if (plugin_name_pattern->match(name.inner)) {
```

Comment:
There is also the option to use the plugin name parser here:
~~~cpp
if (parsers::plugin_name(name.inner)) {
  ...
}
~~~

Reply:
That's better, thanks